### ISSUE TITLE
Add gesture functions

### ISSUE THREAD

**Issue by dpgeorge** [bbcmicrobit/micropython#9]
We should add a simple API for accelerometer gestures, such as: shake, falling, logo_up, logo_down, facing_up, facing_down.

All of them just require sampling the accelerometer over about 1 second and looking at the data to see if the gesture occurred.  That can be done asynchronously and transparent to the user, and just
accessed using simple functions.

We can make a gesture module, or have it as an object as part of the microbit module.  Ie:
```python
import gesture
print(gesture.falling())
# or
import microbit
print(microbit.gesture.falling())
```

There are some subtleties with the API: logo_up, logo_down, facing_up and facing_down are all "continuous" events (like a button being held down) so when you call one of these function it can return True/False if that gesture is currently occurring.

But you might miss a shake or a fall and want methods like fell and shook (or did_fall, did_shake) which return True if the device fell/shook since the last call to these functions, and then return False until the device has the gesture again.  So they are not simple "read only" gestures, they are read&modify.

Is it enough to have falling and shaking and just make sure your code doesn't miss them, or do we also want did_fall and did_shake?

**Comment by ntoll** [comment]
@finneyj any movement on this? We'd like this in our 0.9/1.0 release and may implement something ourselves in a `gestures` module.

**Comment by finneyj** [comment]
apologies - just not had time to do it. I did promise you this though - when's your planned release date?

**Comment by ntoll** [comment]
Tomorrow ;-) (But we're not going to make it.)

Basically, we'll need it pretty damn soon. @dpgeorge had some thoughts about how to implement it.

**Comment by finneyj** [comment]
Folks, please see ```basic-gestures``` branch of microbit-dal. This is as quick as I could manage i'm afraid. Wrote and tested this through today for you...

Looks pretty good. I'm happy to merge this into master, but would appreciate any thoughts you folks have in advanced. Essentially this will give you:

    - TILT_UP
    - TILT_DOWN
    - TILT_LEFT
    - TILT_RIGHT
    - FACE_UP
    - FACE_DOWN
    - SHAKE
    - FREEFALL
    - WHEEE (>=3g)
    - SICK (>=5g)
    - UNCONSCIOUS (>=8g)

*please* do shout loudly if you guys have deadlines - I get pulled in many directions, but if you can let me know of deadlines you have, I'll try to prioritise!

**Comment by finneyj** [comment]
This version has been live on the TD test for a week or two now, with no issues picked up there. Is this going to help you guys in this format? If so, i'll merge into microbit-dal master along with some of the bugfixes you guys have been helping out with shortly... 

Is this also good enough for you guys to be going on with? Anything you guys want to add in advance?

**Comment by markshannon** [comment]
Is there a PR for this, so we can look at the code?

**Comment by dpgeorge** [comment]
@markshannon see here for the commit: https://github.com/lancaster-university/microbit-dal/commit/1b18ac864136cd1c4e01cc555b2606915ebbfeba

**Comment by finneyj** [comment]
sorry - check out basic-gestures branch of microbit-dal.

**Comment by markshannon** [comment]
A PR would make it possible to review the code.
I assume you want it reviewed.

**Comment by finneyj** [comment]
Hi @markshannon 

Sure:. https://github.com/lancaster-university/microbit-dal/pull/65

I've spoken to the MSR guys and this has been live/tested on test.microbit.co.uk for a few weeks now, so largely I wanted to solicit feedback from the micro:python community too, in case you'd like to suggest anything that might improve it form your perspective.

I'm already considering breaking it out into a separate gestures module, but as most gestures are Accelerometer based its slightly more RAM efficient to leave it where it is for now. Might do this later if I find time to bring gestures on board from other sensors (e.g magnetometer).

**Comment by dpgeorge** [comment]
Ok, now that gestures are in the DAL we need to decide how to expose them at the Python level.

The DAL adds a single function called accelerometer.getGesture() which returns the current gesture, one of: NONE, UP, DOWN, LEFT, RIGHT, FACE_UP, FACE_DOWN, FREEFALL, 3G, 6G, 8G, SHAKE.

Mapping directly from the DAL to Python would be something like: microbit.accelerometer.get_gesture().  At the very least that is consistent with get_x(), get_y(), get_z() and get_values().

But would it return an integer code (and then we'd need constants like accelerometer.FACE_UP), or would it return a string?

Do we want to be able to check for gestures that did happen in the past (like button.was_pressed)?  If so we'd need something like was_gesture('FACE_UP').

The original idea (see above) was to have each gesture as a function (eg get_face_up()) in a gesture module.  But it might be best to mirror the DAL and just have a function within the accelerometer.

**Comment by whaleygeek** [comment]
My experience of using gestures in Touch Develop, is that to get good
consistency and performance, you need a 'was up' style. I always implement
these by hand in TD (look at how perfect pancakes does it
www.microbit.co.uk/riotng in the GetGesture function) like this:

on button pressed (A)
    A was pressed := true
end

on shake
   was shake := true
end

on logo down
   was logo down := true
end

It means that your main loop can be relatively slow and in some places
blocking, and you never miss anything important.


___________________________________________________________
David Whale, B.Sc (Hons), MIET
*Software Engineer and IET Schools Liaison Officer, Essex*

email:  dwhale@theiet.org
twitter: @whaleygeek
blog:  blog.whaleygeek.co.uk

Co-author of global bestselling book "Adventures in Minecraft"
<http://amzn.to/ZGfxZG> - lets get kids coding!
Now in English, Spanish, Chinese and Russian.

**Comment by whaleygeek** [comment]
P.S. Sorry for the email footer appearing in the message, I'll endeavour to
post updates from the github webui in the future to prevent this from
happening again.

**Comment by dpgeorge** [comment]
Thanks @whaleygeek for the input.  I agree we need this "was" style, like the buttons already have (and uPy doesn't have callback events so it's not possible to synthesise it like you do in TD).

(BTW, if you reply by email you can just delete the quote of the message you're replying to so it doesn't add the footer.)

**Comment by ntoll** [comment]
I think we should mirror the DAL as @dpgeorge suggests. However, I wonder about it being in its own module? I also believe that returning a string is likely to be more understandable to kids. As a rough and ready naive sketch of the resulting API, we'd perhaps get something like this REPL session:

```
>>> import gestures
>>> gestures.get_gesture()  # Get what's currently happening
"face_up"
>>> gestures.was_gestured()  # Flag to indicate a gesture was detected since the last check.
True
>>> gestures.last_gesture()  # What was the very last gesture detected?
"face_up"
>>> gestures.was_up()  # Convenience function for each of the DAL gestures.
True
>>> gestures.was_up()  # Reset after each call (so this second call is False).
False
```

Does this make sense. It's a straw man so please give it a kick...! I'm not precious about any of the above, but reserve the right to play devil's advocate. ;-)

**Comment by dpgeorge** [comment]
Does was_gestured reset the state or does last_gesture do it?  I think it's confusing if was_up resets but was_gestured does not.  But then what's the point of was_gestured if you can't then call last_gesture straight after it?

**Comment by ntoll** [comment]
Good point. Upon reflection, drop `was_gestured`. It's possible to work out if a gesture has taken place because `gesture.last_gesture` will evaluate to `True`. Good spot! :-)

**Comment by ntoll** [comment]
Hmm... upon even further consideration, I think this is all a bit confusing. I'm going to sleep on it before making further suggestions. I guess with just `last_gesture` we also need an explicit `gestures.clear` too (or not - I've not decided if I like this).

**Comment by dpgeorge** [comment]
The simplest way would be to stick as close to the DAL as possible, and also mirror how we handle the buttons, namely:
```python
microbit.accelerometer.is_gesture('gesture name') # return instantaneous status
microbit.accelerometer.was_gesture('gesture name') # get queued gesture and reset
microbit.accelerometer.get_gestures('gesture name') # get number of gestures of given name
microbit.accelerometer.reset_gestures('gesture name') # reset number for given name
```
If we want a separate module then we could also add convenience functions for each gesture, eg:
```python
gesture.is_up()
gesture.was_up()
gesture.get_ups()
gesture.reset_ups()
```

**Comment by ntoll** [comment]
@dpgeorge wonderful. A lot simpler and far superior to my suggestion. :-)

**Comment by dpgeorge** [comment]
Gestures are now implemented by #160.
